src/monitor.jl

```diff
@@ -0,0 +1,78 @@
+abstract type AbstractMonitor end
+
+using Base: nothing_sentinel
+struct NullMonitor <: AbstractMonitor end
+
+on_update_start!(::AbstractMonitor, inputnames) = nothing
+on_update_node!(::AbstractMonitor, name) = nothing
+on_update_stop!(::AbstractMonitor) = nothing
+
+struct NodeStatistic
+    name::Symbol
+    id::Int64
+    elapsed_time::Int64
+    bytes_allocated::Int
+end
+
+struct RootStatistic
+    name::Symbol
+    id::Int64
+end
+
+mutable struct PerformanceMonitor <: AbstractMonitor
+    @tryconst nodestatistics::Vector{NodeStatistic}
+    @tryconst rootstatistics::Vector{RootStatistic}
+    currentid::Int64
+    lasttime::UInt64
+    @tryconst total_bytes_allocated::Base.RefValue{Int64}
+end
+
+PerformanceMonitor() = PerformanceMonitor(
+    NodeStatistic[],
+    RootStatistic[],
+    zero(Int64),
+    zero(UInt64),
+    Ref(zero(Int64)),
+)
+
+gettime(::PerformanceMonitor) = time_ns()
+getelapsedtime(pm::PerformanceMonitor) = reinterpret(Int64, gettime(pm) - pm.lasttime)
+settime!(pm::PerformanceMonitor) = pm.lasttime = gettime(pm)
+
+setallocatedbytes!(pm::PerformanceMonitor) = Base.gc_bytes(pm.total_bytes_allocated)
+function getallocatedbytes!(pm::PerformanceMonitor)
+    old_total_bytes_allocated = pm.total_bytes_allocated[]
+    Base.gc_bytes(pm.total_bytes_allocated)
+    pm.total_bytes_allocated[] - old_total_bytes_allocated
+    0
```

Comment:
What's this zero doing here?  Seems to me like it shouldn't be here?

Reply:
lol, if you want good performance, you can always learn from [Volkswagen](https://en.wikipedia.org/wiki/Volkswagen_emissions_scandal)...
But yeah, I will fix this mistake! Good spot.